src/test/java/com/puppycrawl/tools/checkstyle/PropertyCacheFileTest.java

```diff
@@ -284,4 +292,200 @@ public void testChangeInConfig() throws Exception {
+                final Configuration config = new DefaultConfiguration("myConfig");
+                final PropertyCacheFile cache = new PropertyCacheFile(config, cacheFile.getPath());
+                cache.load();
+
+                configHashes[i] = cache.get(PropertyCacheFile.CONFIG_HASH_KEY);
+                assertWithMessage("Config hash key should not be null")
+                        .that(configHashes[i])
+                        .isNotNull();
+
+                final Set<String> nonExistentExternalResources = new HashSet<>();
+                final String externalResourceFileName = "non_existent_file.xml";
+                nonExistentExternalResources.add(externalResourceFileName);
+                cache.putExternalResources(nonExistentExternalResources);
+
+                externalResourceHashes[i] = cache.get(PropertyCacheFile.EXTERNAL_RESOURCE_KEY_PREFIX
+                        + externalResourceFileName);
+                assertWithMessage("External resource hashes should not be null")
+                        .that(externalResourceHashes[i])
+                        .isNotNull();
+
+                cache.persist();
+
+                final Properties cacheDetails = new Properties();
+                try (BufferedReader reader = Files.newBufferedReader(cacheFile.toPath())) {
+                    cacheDetails.load(reader);
+                }
+
+                final int expectedNumberOfObjectsInCacheFile = 2;
+                assertWithMessage("Unexpected number of objects in cache")
+                        .that(cacheDetails)
+                        .hasSize(expectedNumberOfObjectsInCacheFile);
+            }
+
+            assertWithMessage("Invalid config hash")
+                    .that(configHashes[0])
+                    .isEqualTo(configHashes[1]);
+            assertWithMessage("Invalid external resource hashes")
+                    .that(externalResourceHashes[0])
+                    .isEqualTo(externalResourceHashes[1]);
+        }
+    }
+
+    /**
+     * It is OK to have long test method name here as it describes the test purpose.
+     */
+    @Test
+    public void testPutNonExistentExternalResourceDifferentExceptionsBetweenRuns()
+            throws Exception {
+        final File cacheFile = File.createTempFile("junit", null, temporaryFolder);
+
+        // We invoke 'putExternalResources' twice to invalidate cache
+        // and have two different exceptions with different content.
+        final int numberOfRuns = 2;
+        final String[] configHashes = new String[numberOfRuns];
+        final String[] externalResourceHashes = new String[numberOfRuns];
+        for (int i = 0; i < numberOfRuns; i++) {
+            final Configuration config = new DefaultConfiguration("myConfig");
+            final PropertyCacheFile cache = new PropertyCacheFile(config, cacheFile.getPath());
+            final CheckstyleException mockException = new CheckstyleException("Exception #" + i);
```

Comment:
do I understand correctly that this test and one above only differ in this exception? It seems to me that it will be better to try to avoid duplication - extract method or use parameterized test

Reply:
We need at least one assert per test, so extracting a method may be the hard way. I'll try to use parametrized test here.

Reply:
Rewritten as a parametrized test.